src/ILLink.RoslynAnalyzer/TrimAnalysis/TrimAnalysisVisitor.cs

```diff
@@ -195,16 +195,15 @@ public override void HandleAssignment (MultiValue source, MultiValue target, IOp
 
 		public override MultiValue HandleArrayElementRead (MultiValue arrayValue, MultiValue indexValue, IOperation operation)
 		{
-			if (arrayValue.AsSingleValue () is not ArrayValue arr)
-				return UnknownValue.Instance;
-
+			MultiValue result = TopValue;
 			if (indexValue.AsConstInt () is not int index)
 				return UnknownValue.Instance;
 
-			if (arr.TryGetValueByIndex (index, out var elementValue))
-				return elementValue;
-
-			return UnknownValue.Instance;
+			foreach (var value in arrayValue) {
```

Comment:
Why do we need to look over multiple `arrayValue`s here but not in `ScanLdelem`? Is it because the analyzer has made a copy of the `ArrayValue` and merged them into something that has multiple arrays (instead of one array with some merged value at the known index?

I think we should aim to have the linker and analyzer use essentially the same logic here. I would be ok with copying the `ArrayValue` and tracking merged values as multiple arrays since it seems like that is already what we do in the analyzer.

One alternative would be to maintain a notion of object identity, and adjust the meet operation so that meeting two array `ArrayValues` which represent the same object in memory, but with different elements, produces a single `ArrayValue` whose elements are the merged values. But I don't think it's worth it at the moment.

Reply:
I looked at the linker code and we are executing `ScanStelem` which is serving the purpose of going through the array and merging the elements together. I don't think I see a lot of difference with the `HandleArrayElementRead`. If something analyzer code might be more efficient because it only asks once for the index being null